### PR TITLE
Resolve imports and exports via cached BindingsMap

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Addresses #6100 a bit. `org.enso.compiler.Compiler.runImportsAndExportsResolution` used to take up to three seconds as it triggered deserialization of megabytes of `IR` caches. Now it is faster as it uses _library bindings cache_. Alas, overall time isn't shortened, as further compiler passes still load (the same amount of) `IR` caches in.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- All code has been tested:
  - [x] Unit tests continue to pass

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
old importsAndExportsResolution

The above picture represents the current state. The picture below shows the after applying this PR. The `runImportsAndExportsResolution` clearly no longer deserializes `IR` - just it loads library bindings - when they are available.

new importsAndExportsResolution

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java [hunk 1]
```diff
@@ -258,17 +258,24 @@ public CompilerContext.Module findTopScopeModule(String name) {
 
 
   private final class ModuleUpdater implements Updater, AutoCloseable {
-
+    private final Module compilerModule;
     private final org.enso.interpreter.runtime.Module module;
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java]: rename this to `runtimeModule`?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java [hunk 2]
```diff
@@ -258,17 +258,24 @@ public CompilerContext.Module findTopScopeModule(String name) {
...
     private Boolean loadedFromCache;
     private Boolean hasCrossModuleLinks;
     private boolean resetScope;
     private boolean invalidateCache;
 
-    private ModuleUpdater(org.enso.interpreter.runtime.Module module) {
-      this.module = module;
+    private ModuleUpdater(Module module) {
+      this.compilerModule = module;
+      this.module = module.module;
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java]: I find this indirection difficult to read. Is it possible to, e.g., remove `this.module` field, and refer to it as, e.g., `this.compilerModule.runtimeModule`?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java [hunk 3]
```diff
@@ -303,6 +310,12 @@ public void invalidateCache() {
 
     @Override
     public void close() {
+      if (map != null) {
+        if (compilerModule.bindings != null) {
+          throw new IllegalStateException("Reassigining bindings to " + compilerModule);
+        }
+        compilerModule.bindings = map;
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java]: So the BindingsMap is not saved to the runtime `module`, but only to `compilerModule`, which is kind of a stub for the runtime module.

File: engine/runtime/src/main/scala/org/enso/compiler/phase/ImportResolver.scala [hunk 4]
```diff
@@ -109,6 +109,13 @@ class ImportResolver(compiler: Compiler) {
                 val converted = bindings
                   .toConcrete(compiler.packageRepository.getModuleMap)
                   .map { concreteBindings =>
+                    compiler.context.updateModule(
```
Comment (Akirathan) [inline engine/runtime/src/main/scala/org/enso/compiler/phase/ImportResolver.scala:112]: So this is the place where `BindingsMap` is set into the module from the cache retrieved by `compiler.importExportBindings(current)`. Nice.

File: engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala [hunk 5]
```diff
@@ -435,9 +436,8 @@ class Compiler(
           val shouldStoreCache =
             irCachingEnabled && !context.wasLoadedFromCache(module)
           if (
-            shouldStoreCache && !hasErrors(module) && !context.isInteractive(
-              module
-            )
+            shouldStoreCache && !hasErrors(module) &&
+            !context.isInteractive(module) && !context.isSynthetic(module)
```
Comment (Akirathan) [inline engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:440]: Why is this change needed? Why skip serialization of synthetic modules? Isn't it possible that this leads to some inconsistencies? Could synthetic modules be generated on the fly during deserialization of other modules?

File: engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala [hunk 6]
```diff
@@ -297,6 +297,7 @@ class Compiler(
     var hasInvalidModuleRelink = false
     if (irCachingEnabled) {
       requiredModules.foreach { module =>
+        ensureParsed(module)
```
Comment (Akirathan) [inline engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:300]: Why is this line added here? Could you pls add a comment? In the above code, I can see that under certain circumstances, there is `parseModule(module)` called. Isn't that enough?

File: engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala [hunk 7]
```diff
@@ -522,11 +524,8 @@ class Compiler(
             .map { concreteBindings =>
               concreteBindings
             }
-          val ir = context.getIr(module)
-          val currentLocal = ir.unsafeGetMetadata(
-            BindingAnalysis,
-            "Synthetic parsed module missing bindings"
-          )
+          ensureParsed(module)
```
Comment (Akirathan) [inline engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:527]: Another `ensureParsed` for a synthetic module? If the serialization of synthetic modules would not be skipped, could this be removed? See [my other comment](https://github.com/enso-org/enso/pull/8160/files#r1374285736).

**Review by Akirathan — APPROVED** [review]
Nice improvement. Approving, although I would appreciate more clarity (more docs).

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java (hunk 2 above)
Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java]: This or use `runtimeModule` name when appropriate. It's really hard to follow which module we refer to

File: engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala (hunk 6 above)
Comment (hubertp) [reply engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:300]: I also don't understand this change. Required module should already be parsed.

File: engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala (hunk 5 above)
Comment (hubertp) [reply engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:440]: > Why skip serialization of synthetic modules? Isn't it possible that this leads to some inconsistencies?

I don't remember the exact reasons but I guess having a mixed scenario of cached and uncached IRs might be one.

File: engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala [hunk 8]
```diff
@@ -95,6 +95,9 @@ final class SerializationManager(compiler: Compiler) {
     useGlobalCacheLocations: Boolean,
     useThreadPool: Boolean = true
   ): Future[Boolean] = {
+    if (module.isSynthetic) {
+      throw new IllegalStateException();
```
Comment (hubertp) [inline engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala]: Please provide some message with it

File: engine/runtime/src/main/scala/org/enso/compiler/phase/ImportResolver.scala [hunk 9]
```diff
@@ -109,6 +109,13 @@ class ImportResolver(compiler: Compiler) {
...
+                      current,
+                      { u =>
+                        u.bindingsMap(concreteBindings)
+                        u.loadedFromCache(true)
```
Comment (hubertp) [inline engine/runtime/src/main/scala/org/enso/compiler/phase/ImportResolver.scala:116]: Should we maybe have more than 2 states? It feels like 1) `not yet loaded` 2) `loaded only bindings` 3) `full IR loaded` would be more appropriate

**Review by hubertp — COMMENTED** [review]
I don't understand the reasons for additional `ensureParsed` calls

File: engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala (hunk 5 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:440]: According to @hubertp, synthetic modules aren't supposed to be serialized (moreover they cannot be serialized, as they yield exceptions during serialization).

File: engine/runtime/src/main/scala/org/enso/compiler/phase/ImportResolver.scala (hunk 9 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/phase/ImportResolver.scala:116]: There is nobody reading the "three states" yet. Why shall we keep three states in memory when we only read two?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java]: Unified in 8c4ef0f588

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java]: Unified in 8c4ef0f588

File: engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala (hunk 6 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:300]: Removing the `parseModule(module)` here leads to:
```
[error] Test org.enso.compiler.SerdeCompilerTest.testFibTest failed: java.lang.NullPointerException: Cannot invoke "org.enso.compiler.core.ir.Module.preorder()" because the return value of "org.enso.compiler.context.CompilerContext.getIr(org.enso.compiler.context.CompilerContext$Module)" is null, took 0.257 sec
[error]     at org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$6(Compiler.scala:303)
[error]     at org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$6$adapted(Compiler.scala:299)
[error]     at scala.collection.immutable.List.foreach(List.scala:333)
[error]     at org.enso.compiler.Compiler.runCompilerPipeline(Compiler.scala:299)
[error]     at org.enso.compiler.Compiler.go$1(Compiler.scala:233)
[error]     at org.enso.compiler.Compiler.runInternal(Compiler.scala:240)
[error]     at org.enso.compiler.Compiler.run(Compiler.scala:136)
[error]     at org.enso.compiler.SerdeCompilerTest.parseSerializedModule(SerdeCompilerTest.java:89)
[error]     at org.enso.compiler.SerdeCompilerTest.testFibTest(SerdeCompilerTest.java:36)
```
why?

Now, when we use `BindingsMap` in the `runImportsAndExportsResolution`, it may happen that modules get to this phase without having `IR` (just the `BindingsMap`). However the `requiredModules` will need `IR` to pass thru `runGlobalTypingPasses` at line 350. E.g. we have to load the `IR` for them. For example at this moment.

File: engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala (hunk 7 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:527]: I have to admit, I am not sure whit call is good for. Things seem to run without it. Let's remove it 5d96d7d6cc5fb00f6793fa04cd9ba63d47b46796 and see how far we can get!

File: engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala (hunk 8 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala]: Done in 7b4ab282ac

File: engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala (hunk 7 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:527]: Probably it wasn't that great idea to remove it. We are getting [failure in Numpy project test](https://github.com/enso-org/enso/actions/runs/6667090478/job/18119889353?pr=8160#step:10:4503) on Linux and Mac. However **locally everything works** and I am not able to simulate the failure. Trying to force [cache cleanup](https://github.com/enso-org/enso/pull/8160/commits/edbec01ff3ad027b48d83f88f5eac9d0e520b7cb).

If that doesn't help, I'll return the `ensureParsed` back - as before 5d96d7d6cc5fb00f6793fa04cd9ba63d47b46796 the test was passing, I think.

Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:527]: After [bcad8e4](https://github.com/enso-org/enso/pull/8160/commits/bcad8e4dba7eb46a147e09730f784bfc7e557bdb) the _CI is  green_ again. I guess that explains why `ensureParsed` is needed here (at least for current moment).
